src/Dialect/Krnl/KrnlHelper.cpp

```diff
@@ -158,6 +158,41 @@ void KrnlIterateOperandPack::pushAffineMapBound(
     _operands.emplace_back(operand);
 }
 
+// Bound could be a constant, Value or AffineMap
+void KrnlIterateOperandPack::pushIndexExprBound(IndexExpr expr) {
+  if (expr.isLiteral()) {
+    pushConstantBound(expr.getLiteral());
+  } else if (expr.isAffine() && !expr.isPredType()) {
+    int dimNum = expr.getContext().getNumDims();
+    int symNum = expr.getContext().getNumSymbols();
+    AffineMap map = AffineMap::get(dimNum, symNum, {expr.getAffineExpr()},
+        expr.getRewriter().getContext());
+    SmallVector<Value, 4> list;
+    expr.getContext().getDimAndSymbolList(list);
+    pushAffineMapBound(map, list);
+  } else {
+    // Assume the expr is loop invariant if there is any outer loop
+    pushOperandBound(expr.getValue());
+  }
+}
+
+void KrnlIterateOperandPack::pushIndexExprsBound(
+    SmallVectorImpl<IndexExpr> &exprVector) {
+  SmallVector<AffineExpr, 4> AEVector;
```

Comment:
I would do the same here, and declare it as a static for class IndexExpr, passing the vector of Index expressions.